### PR TITLE
S2 data retention

### PR DESCRIPTION
**Pull request by srikanthreddy5**
when "remotePath = volume:remote_store/$_index_name"  and _maxDailyDataSizeMB used in indexes databag it used to calculate "maxGlobalDataSizeMB" . 

https://docs.splunk.com/Documentation/Splunk/7.3.3/Indexer/SmartStoredataretention

### CODE REVIEW

File: recipes/_configure_indexes.rb [hunk 1]
```diff
@@ -51,8 +52,12 @@
 
     padding ||= default_config['_dataSizePaddingPercent']
     padding = padding.nil? ? 1.1 : 1 + (padding / 100.0)
-
-    hash['maxTotalDataSizeMB'] = (daily_mb * padding * frozen_time_in_days * replication_factor).to_i / indexer_count
+    # For smartstore enabled indexes only .
+    if s2_enabled_index == true
+      hash['maxGlobalDataSizeMB'] = (daily_mb * padding * frozen_time_in_days * replication_factor).to_i / indexer_count
```
Comment (gravesb) [inline recipes/_configure_indexes.rb:57]: This calculation should be different since this is a global value instead of per indexer. I don't think the rep factor or indexer count should be relevant here.

File: spec/unit/recipes/_configure_indexes_spec.rb [hunk 2]
```diff
@@ -138,33 +138,44 @@
     end
   end
 
-  context 'when _noGenerateTstatsHomePath is set to true for specific indexes' do
+  context 'when _noGenerateTstatsHomePath is set to true and _is_s2Index is set to true for specific indexes' do
```
Comment (gravesb) [inline spec/unit/recipes/_configure_indexes_spec.rb:141]: rather than conflating these contexts in the same tests, I would prefer to have separate contexts to test the new settings.

**Review by gravesb — CHANGES_REQUESTED** [review]

File: spec/unit/recipes/_configure_indexes_spec.rb (hunk 2 above)
Comment (srikanthreddy5) [reply spec/unit/recipes/_configure_indexes_spec.rb]: I have created separate test now for [rspec_tests](https://github.com/cerner/cerner_splunk/pull/241/commits/eb38d4732d72ac6a0ce03a1c28f55e5591cc2868#diff-2dbf75fe595324af1923d6c51037b9e78cab1adcc9d842af08ba7e0efba026d4R226-R306)

File: recipes/_configure_indexes.rb [hunk 3]
```diff
@@ -39,9 +39,10 @@
   daily_mb = hash.delete('_maxDailyDataSizeMB')
   padding = hash.delete('_dataSizePaddingPercent')
   default_config = config.fetch('default', {})
+  s2_enabled_index = hash.delete('_is_s2Index') || default_config['_is_s2Index']
```
Comment (gravesb) [inline recipes/_configure_indexes.rb:42]: Can't we trigger off of either the remotePath setting or the storageType=remote configurations instead?  Those should already be provided in the databag to enable smartstore, so we shouldn't need a separate specific flag for it.

File: recipes/_configure_indexes.rb [hunk 4]
```diff
@@ -39,9 +39,10 @@
...
   # _noGenerateTstatsHomePath is false  by default.
   no_gentstat = hash.delete('_noGenerateTstatsHomePath') || default_config['_noGenerateTstatsHomePath']
-  if %i[index default].include?(stanza_type) && daily_mb && !hash.key?('maxTotalDataSizeMB')
+  if %i[index default].include?(stanza_type) && daily_mb && !hash.key?('maxTotalDataSizeMB') && !hash.key?('maxGlobalDataSizeMB')
```
Comment (gravesb) [inline recipes/_configure_indexes.rb:45]: This allows for a potentially unexpected interaction.  If they set the maxTotalDataSize and enable s2, but don't set the maxGlobalDataSize then the global data size won't be auto generated.  It's not something that should ever happen, but if it does I think we could handle it better.  Possibly something like:
`%i[index default].include?(stanza_type) && daily_mb && ((!s2_enabled_index && !hash.key?('maxTotalDataSizeMB')) || (s2_enabled_index && !hash.key?('maxGlobalDataSizeMB')))` though there may be a better way.

**Review by gravesb — COMMENTED** [review]
Please also delete the empty `chef:` file.

File: recipes/_configure_indexes.rb (hunk 3 above)
Comment (srikanthreddy5) [reply recipes/_configure_indexes.rb:42]: Smart-store configuration for indexes.conf is deployed using an App, so remotePath and storageType=remote are not configured in data bag.

Comment (gravesb) [reply recipes/_configure_indexes.rb:42]: So change where you configure it, or just add storageType=remote to the databag as well. The main point here is there are existing configurations for Splunk to indicate a remote store index.  We should use those to determine remote store rather than inventing custom configurations for it.

File: spec/unit/recipes/_configure_indexes_spec.rb [hunk 5]
```diff
@@ -222,4 +224,88 @@
...
+            'maxGlobalDataSizeMB' => 60_060
+          }
+        }
+      }
+
+      expect(subject).to create_splunk_template('system/indexes.conf').with(expected_attributes)
+    end
+  end
+
+  context 'when remotePath is set to true in specific stanza' do
```
Comment (gravesb) [inline spec/unit/recipes/_configure_indexes_spec.rb:270]: minor, remotePath isn't a boolean.  Reword this to something like 'when remotePath is set in specific stanza'.

File: recipes/_configure_indexes.rb [hunk 6]
```diff
@@ -39,9 +39,11 @@
   daily_mb = hash.delete('_maxDailyDataSizeMB')
   padding = hash.delete('_dataSizePaddingPercent')
   default_config = config.fetch('default', {})
+
+  s2_enabled_index =  hash.key?('remotePath') || !default_config['remotePath'].nil?
```
Comment (gravesb) [inline recipes/_configure_indexes.rb:43]: minor, we're doing the same check in two different ways here. To make the code a bit easier to read, please update them to be the same.  Either use .key? on both or check for ! .nil? on both.

**Review by gravesb — CHANGES_REQUESTED** [review]

File: recipes/_configure_indexes.rb (hunk 6 above)
Comment (srikanthreddy5) [reply recipes/_configure_indexes.rb]: updated to check for remotePath [43ed5fe](https://github.com/cerner/cerner_splunk/pull/241/commits/43ed5fe18075cff99d99acbdcc302e2427146dbb#diff-73c916ec93eae5658af563fbf3f9eedd6ad279c110b632537704da1014ce0c1dR43)

File: spec/unit/recipes/_configure_indexes_spec.rb (hunk 5 above)
Comment (srikanthreddy5) [reply spec/unit/recipes/_configure_indexes_spec.rb]: updated [43ed5fe](https://github.com/cerner/cerner_splunk/pull/241/commits/43ed5fe18075cff99d99acbdcc302e2427146dbb#diff-2dbf75fe595324af1923d6c51037b9e78cab1adcc9d842af08ba7e0efba026d4R314)

File: spec/unit/recipes/_configure_indexes_spec.rb [hunk 7]
```diff
@@ -179,12 +180,13 @@
     end
   end
 
-  context 'when _noGenerateTstatsHomePath is set to true in default stanza' do
+  context 'when _noGenerateTstatsHomePath is set to true or remotePath is set in default stanza' do
```
[the remark this answers is not in the document]
Comment (srikanthreddy5) [reply spec/unit/recipes/_configure_indexes_spec.rb]: removed remotePath is set from context [43ed5fe](https://github.com/cerner/cerner_splunk/pull/241/commits/43ed5fe18075cff99d99acbdcc302e2427146dbb#diff-2dbf75fe595324af1923d6c51037b9e78cab1adcc9d842af08ba7e0efba026d4R183)

File: recipes/_configure_indexes.rb [hunk 8]
```diff
@@ -63,7 +69,7 @@
       hash['coldPath'] = "#{base_path}/#{dir_name}/colddb" unless hash['coldPath']
       hash['homePath'] = "#{base_path}/#{dir_name}/db" unless hash['homePath']
       hash['thawedPath'] = "$SPLUNK_DB/#{dir_name}/thaweddb" unless hash['thawedPath']
-      hash['tstatsHomePath'] = "#{base_path}/#{dir_name}/datamodel_summary" if volume && !hash['tstatsHomePath'] && !no_gentstat
+      hash['tstatsHomePath'] = "#{base_path}/#{dir_name}/datamodel_summary" if volume && !hash['tstatsHomePath'] && !no_gentstat && !s2_enabled_index
```
[the remark this answers is not in the document]
Comment (srikanthreddy5) [reply recipes/_configure_indexes.rb]: removed  "!s2_enabled_index" [43ed5fe](https://github.com/cerner/cerner_splunk/pull/241/commits/43ed5fe18075cff99d99acbdcc302e2427146dbb#diff-73c916ec93eae5658af563fbf3f9eedd6ad279c110b632537704da1014ce0c1dR72)

File: recipes/_configure_indexes.rb [hunk 9]
```diff
@@ -51,8 +53,12 @@
 
     padding ||= default_config['_dataSizePaddingPercent']
     padding = padding.nil? ? 1.1 : 1 + (padding / 100.0)
-
-    hash['maxTotalDataSizeMB'] = (daily_mb * padding * frozen_time_in_days * replication_factor).to_i / indexer_count
+    # For smartstore enabled indexes only .
+    if s2_enabled_index == true
```
[the remark this answers is not in the document]
Comment (srikanthreddy5) [reply recipes/_configure_indexes.rb]: Removed _== true_ [43ed5fe](https://github.com/cerner/cerner_splunk/pull/241/commits/43ed5fe18075cff99d99acbdcc302e2427146dbb#diff-73c916ec93eae5658af563fbf3f9eedd6ad279c110b632537704da1014ce0c1dR57)

File: docs/databags.md [hunk 10]
```diff
@@ -78,9 +78,10 @@ An Indexes Hash is part of a plaintext data bag item that defines the set of ind
...
-    * `_maxDailyDataSizeMB` - The amount of daily usage this index is expected to consume.  Used to calculate the maxTotalDataSizeMB if maxTotalDataSizeMB has not already been specified for the index.
+    * `_maxDailyDataSizeMB` - The amount of daily usage this index is expected to consume. Used to calculate the maxTotalDataSizeMB or maxGlobalDataSizeMB
+    if maxTotalDataSizeMB or maxGlobalDataSizeMB has not already been specified for the index.
     * `_dataSizePaddingPercent` - The percentage of padding to apply to the amount of space this index is expected to consume.  Used to calculate the maxTotalDataSizeMB if maxTotalDataSizeMB has not already been specified for the index. Defaults to 10 if no value is specified.
-    * `_noGenerateTstatsHomePath` - Do not generate tstatsHomePath for this index when set to true. Used to disable datamodel acceleration in case of the smartstore enabled indexes. 
+    * `_noGenerateTstatsHomePath` - Do not generate tstatsHomePath for this index when set to true. Used to disable datamodel acceleration in case of the smartstore enabled indexes.This is depricated, instead of this please set the remotePath in smartstore enabled clusters.
```
[the remark this answers is not in the document]
Comment (srikanthreddy5) [reply docs/databags.md]: updated misspelled word [43ed5fe](https://github.com/cerner/cerner_splunk/pull/241/commits/43ed5fe18075cff99d99acbdcc302e2427146dbb#diff-424d7b6dcf972936e356bc5be64f0b83bf7336701a8b30fcb4df26bb9b995cacR84)

File: spec/unit/recipes/_configure_indexes_spec.rb [hunk 11]
```diff
@@ -222,4 +268,88 @@
       expect(subject).to create_splunk_template('system/indexes.conf').with(expected_attributes)
     end
   end
+
+  context 'when remotePath is set in default stanza' do
```
Comment (gravesb) [inline spec/unit/recipes/_configure_indexes_spec.rb:272]: minor, "when remotePath is set in default stanza and _maxDailyDataSizeMB is given"

File: spec/unit/recipes/_configure_indexes_spec.rb [hunk 12]
```diff
@@ -222,4 +268,88 @@
...
+            'maxGlobalDataSizeMB' => 60_060
+          }
+        }
+      }
+
+      expect(subject).to create_splunk_template('system/indexes.conf').with(expected_attributes)
+    end
+  end
+
+  context 'when remotePath is set in specific stanza' do
```
Comment (gravesb) [inline spec/unit/recipes/_configure_indexes_spec.rb:314]: minor, "when remotePath is set in specific stanza and _maxDailyDataSizeMB is given"

**Review by gravesb — APPROVED** [review]
A few minor comments on spec contexts, but otherwise I think this looks good.

File: spec/unit/recipes/_configure_indexes_spec.rb (hunk 11 above)
Comment (srikanthreddy5) [reply spec/unit/recipes/_configure_indexes_spec.rb]: updated, Thanks! [273d89b](https://github.com/cerner/cerner_splunk/pull/241/commits/273d89beef2330782b1fa1e4360b0ab359130995#diff-2dbf75fe595324af1923d6c51037b9e78cab1adcc9d842af08ba7e0efba026d4R272)

File: spec/unit/recipes/_configure_indexes_spec.rb (hunk 12 above)
Comment (srikanthreddy5) [reply spec/unit/recipes/_configure_indexes_spec.rb]: updated, Thanks! [273d89b](https://github.com/cerner/cerner_splunk/pull/241/commits/273d89beef2330782b1fa1e4360b0ab359130995#diff-2dbf75fe595324af1923d6c51037b9e78cab1adcc9d842af08ba7e0efba026d4R314)

**Review by som3guy — APPROVED** [review]

**Comment by gravesb** [comment]
Merged manually via 34e9a5a
